### PR TITLE
black-/whitelists

### PR DESCRIPTION
**Pull request by robinkanters**

### CODE REVIEW

**Comment by zapline** [comment]
useful

**Comment by robinkanters** [comment]
> op-list also need

I made it so that that can very easily be added: override `PlayerList`, give it a `ConfigFileName` and check it with something like

    private static bool PlayerIsOp(IRemoteClient client, IMultiplayerServer server)
    {
        if (!_oplistLoaded)
        {
            var oplist = new PlayerOpList(server);
            _oplistedPlayers = oplist.Players;
            _oplistLoaded = true;
        }
        return _oplistedPlayers != null && _oplistedPlayers.Contains(client.Username, StringComparer.CurrentCultureIgnoreCase);
    }

**Comment by SirCmpwn** [comment]
I would prefer to manage the black/whitelist out of config.yaml.

**Comment by robinkanters** [comment]
I can see why, however I'm concerned that bigger servers might get long lists and therefore long config files... Maybe implement both and make it configurable?

**Comment by SirCmpwn** [comment]
Hmm, good point. I think we should go with `whitelist/blacklist.yaml`, then (*not* `.txt`).

**Comment by robinkanters** [comment]
I'm gonna let this sit in the back of my head for a while and come back to it :blush:

**Comment by robinkanters** [comment]
I'm wondering how you'd like to see the structure of a 'blacklist.yml' file...

Something like this?

    - name1
    - name2
    - name3

Because I'd prefer something like

    - blacklist
      - name1
      - name2

But that might give the impression that you can add a `whitelist

**Comment by SirCmpwn** [comment]
- blacklist
        - ips
            - 192.30.252.0/22
            - 204.232.175.64
        - users
            - SirCmpwn
            - notch
            - Grum
    - whitelist
        - ips
            - 192.30.252.20
        - users
            - Dinnerbone

If we add support for wildcards, you can make an invite-only server:

    - blacklist
        - users
            - *
    - whitelist
        - users
            - SirCmpwn
            - notch
            - Grum

**Comment by robinkanters** [comment]
That would put everything in one file. Which is fine, but would require a bit of an overhaul. In that case, I'd like to rename it to something like `security.yml` and have that for potential future security settings

**Comment by SirCmpwn** [comment]
How about `access.yaml`?

**Comment by robinkanters** [comment]
I see the concept of 'access' as a subset of the concept of 'security', so I prefer 'security' to leave room for future additions. Also, @zapline proposed an `oplist` as well, which would makes sense under `security`, but not under `access` IMHO

**Comment by SirCmpwn** [comment]
`access` works for ops, I think. The idea is that it would define different levels of access to the server - be that none (blacklisted), some (whitelisted), or all (op'd).

**Comment by robinkanters** [comment]
Then still my first point

**Comment by robinkanters** [comment]
I'm wondering what you don't like about `security`

**Comment by SirCmpwn** [comment]
I feel like `security` things beyond black/whitelists would go into `config.yaml`. For example, `enableSSL: yes`. Also, I would guess that `access.yaml` is going to be edited by code and `config.yaml` won't be.

**Comment by robinkanters** [comment]
I just feel like access not only applies to black/white/oplists, but also the connection to the server. At least, I'd probably expect that as a user. I'm going to see if I can come up with a better name :relaxed: . I'm probably not programming a lot this weekend, anyway.

By the way,

> access.yaml is going to be edited by code

yeah, nice feature, but that's not implemented yet

**Comment by SirCmpwn** [comment]
>yeah, nice feature, but that's not implemented yet

Sure, I'm talking about the future. `/ban`, `/op`, etc.

**Comment by SirCmpwn** [comment]
Now you have a shitload of commits for one feature, can you rebase?

**Comment by robinkanters** [comment]
Not now, heading to work. May take a look tonight

**Comment by SirCmpwn** [comment]
I didn't expect it to be taken care of immediately. Take your time.

**Comment by robinkanters** [comment]
that went easier than expected

File: TrueCraft/MultiplayerServer.cs [hunk 1]
```diff
@@ -193,7 +198,7 @@ public IEntityManager GetEntityManagerForWorld(IWorld world)
 
         public void SendMessage(string message, params object[] parameters)
         {
-            var compiled = string.Format(message, parameters);
+            var compiled = String.Format(message, parameters);
```
Comment (SirCmpwn) [inline TrueCraft/MultiplayerServer.cs]: Why did you replace `string` with `String` everywhere?

Comment (robinkanters) [inline TrueCraft/MultiplayerServer.cs]: Blame visual studio ;) slipped past my sanity check, will fix after dinner

Comment (SirCmpwn) [inline TrueCraft/MultiplayerServer.cs]: It's the only issue here, I'll just correct it myself during the merge. Thanks!

**Comment by robinkanters** [comment]
Okay great ;)

**Pull request merged**